hail/python/hail/vds/combiner/variant_dataset_combiner.py

```diff
@@ -612,20 +615,46 @@ def maybe_load_from_saved_path(save_path: str) -> Optional[VariantDatasetCombine
     if isinstance(reference_genome, str):
         reference_genome = hl.get_reference(reference_genome)
 
+    # we need to compute the type that the combiner will have, this will allow us to read matrix
+    # tables quickly, especially in an asynchronous environment like query on batch where typing
+    # a read uses a blocking round trip.
+    gvcf_type = None
     if gvcf_reference_entry_fields_to_keep is None and vds_paths:
         vds = hl.vds.read_vds(vds_paths[0])
-        gvcf_reference_entry_fields_to_keep = set(vds.reference_data.entry) - {'END'}
+        if gvcf_reference_entry_fields_to_keep is None:
+            gvcf_reference_entry_fields_to_keep = set(vds.reference_data.entry) - {'END'}
+        dataset_type = CombinerOutType(reference_type=vds.reference_data._type, variant_type=vds.variant_data._type)
     elif gvcf_reference_entry_fields_to_keep is None and gvcf_paths:
-        mt = hl.import_vcf(gvcf_paths[0], force_bgz=True, reference_genome=reference_genome,
+        mt = hl.import_vcf(gvcf_paths[0], header_file=gvcf_external_header, force_bgz=True,
+                           array_elements_required=False, reference_genome=reference_genome,
+                           contig_recoding=contig_recoding)
+        gvcf_type = mt._type
+        rmt = mt.filter_rows(hl.is_defined(mt.info.END))
+        gvcf_reference_entry_fields_to_keep = defined_entry_fields(rmt, 100_000) - {'GT', 'PGT', 'PL'}
+
+        vds = transform_gvcf(mt._key_rows_by_assert_sorted('locus'), gvcf_reference_entry_fields_to_keep, gvcf_info_to_keep)
+        dataset_type = CombinerOutType(reference_type=vds.reference_data._type, variant_type=vds.variant_data._type)
+    else:
+        mt = hl.import_vcf(gvcf_paths[0], header_file=gvcf_external_header, force_bgz=True,
+                           array_elements_required=False, reference_genome=reference_genome,
                            contig_recoding=contig_recoding)
-        mt = mt.filter_rows(hl.is_defined(mt.info.END))
-        gvcf_reference_entry_fields_to_keep = defined_entry_fields(mt, 100_000) - {'GT', 'PGT', 'PL'}
+        gvcf_type = mt._type
+        vds = transform_gvcf(mt._key_rows_by_assert_sorted('locus'), gvcf_reference_entry_fields_to_keep, gvcf_info_to_keep)
+        dataset_type = CombinerOutType(reference_type=vds.reference_data._type, variant_type=vds.variant_data._type)
+    if gvcf_type is None and gvcf_paths:
+        mt = hl.import_vcf(gvcf_paths[0], header_file=gvcf_external_header, force_bgz=True,
+                           array_elements_required=False, reference_genome=reference_genome,
+                           contig_recoding=contig_recoding)
+        gvcf_type = mt._type
```

Comment:
I did some fiddling with this and I think the following is both correct and easier to reason about. What do you think of it? It still doesn't ensure that the extant VDS type matches the GVCF type. I presume that'll fail later?

```python3
    vds = None
    gvcf_paths = None
    if vds_paths:
        vds = hl.vds.read_vds(vds_paths[0])
        if gvcf_reference_entry_fields_to_keep is None:
            gvcf_reference_entry_fields_to_keep = set(vds.reference_data.entry) - {'END'}
    if gvcf_paths:
        gvcf_mt = hl.import_vcf(
            gvcf_paths[0],
            header_file=gvcf_external_header,
            force_bgz=True,
            array_elements_required=False,
            reference_genome=reference_genome,
            contig_recoding=contig_recoding
        )
        gvcf_type = gvcf_mt._type
        if gvcf_reference_entry_fields_to_keep is None:
            rmt = gvcf_mt.filter_rows(hl.is_defined(gvcf_mt.info.END))
            gvcf_reference_entry_fields_to_keep = defined_entry_fields(rmt, 100_000) - {'GT', 'PGT', 'PL'}
        if vds is None:
            vds = transform_gvcf(
                gvcf_mt._key_rows_by_assert_sorted('locus'),
                gvcf_reference_entry_fields_to_keep,
                gvcf_info_to_keep
            )

    dataset_type = CombinerOutType(
        reference_type=vds.reference_data._type,
        variant_type=vds.variant_data._type
    )
```